File: src/main/java/com/jcloisterzone/ai/player/RankingAiPlayer.java
package com.jcloisterzone.ai.player;

import java.io.InputStream;
import java.io.PrintStream;
import java.util.Scanner;
import java.util.stream.Collectors;

import com.jcloisterzone.ai.AiPlayer;
import com.jcloisterzone.ai.GameStateRanking;
import com.jcloisterzone.Player;
import com.jcloisterzone.game.GameSetup;
import com.jcloisterzone.game.GameStatePhaseReducer;
import com.jcloisterzone.game.capability.PortalCapability;
import com.jcloisterzone.game.state.GameState;
import com.jcloisterzone.io.message.PlaceTileMessage;
import com.jcloisterzone.io.message.ReplayableMessage;

import io.vavr.Tuple2;
import io.vavr.collection.Queue;
import io.vavr.collection.Vector;

public abstract class RankingAiPlayer implements AiPlayer {

    private GameStateRanking stateRanking;
    private GameStatePhaseReducer phaseReducer;

    private Player me;
    private Vector<ReplayableMessage> messages = Vector.empty();

    /**
     * Instantiates a new {@code PlayerAction}.
     *
     * @param options the options the player can choose from
     */
    public RankingAiPlayer(GameStatePhaseReducer phaseReducer, Player me) {
       this.phaseReducer = phaseReducer;
       this.me = me;
       this.stateRanking = this.createStateRanking(me);
    }
    protected abstract GameStateRanking createStateRanking(Player me);

    @Override
    public ReplayableMessage apply(GameState state) {
        if (messages.isEmpty()) {
            Double bestSoFar = Double.NEGATIVE_INFINITY;
            Queue<Tuple2<GameState, Vector<ReplayableMessage>>> queue = Queue.of(new Tuple2<>(state, Vector.empty()));

            while (!queue.isEmpty()) {
                Tuple2<Tuple2<GameState, Vector<ReplayableMessage>>, Queue<Tuple2<GameState, Vector<ReplayableMessage>>>> t = queue.dequeue();
                queue = t._2;
                Tuple2<GameState, Vector<ReplayableMessage>> item = t._1;
                GameState itemState = item._1;

                for (ReplayableMessage msg : getPossibleActions(itemState)) {
                    Vector<ReplayableMessage> chain = item._2.append(msg);
                    GameState newState = phaseReducer.apply(itemState, msg);
                    boolean end = newState.getActivePlayer() != me || newState.getTurnPlayer() != state.getTurnPlayer();

                    if (!end && msg instanceof PlaceTileMessage &&
                        newState.getLastPlaced().getTile().hasModifier(PortalCapability.MAGIC_PORTAL)) {
                        // hack to avoid bad performance on Portal tile
                        // rank just placement then rang meeple placement separately
                        // still not perfect because it can miss good on tile meeple placement
                        end = true;
                    }

                    if (end) {
                        Double ranking = stateRanking.apply(newState);

//                      String chainStr = chain.map(_msg -> _msg.getClass().getSimpleName()).toJavaStream().collect(Collectors.joining(", "));
//                      System.err.println(String.format(">>> %f\n%s", ranking, chainStr));

                        if (ranking > bestSoFar) {
                            bestSoFar = ranking;
                            messages = chain;
                        }
                    } else {
                        queue = queue.enqueue(new Tuple2<>(newState, chain));
                    }
                }
            }

            String chainStr = messages.map(_msg -> _msg.getClass().getSimpleName()).toJavaStream().collect(Collectors.joining(", "));
            System.out.println(String.format("Best ranking %s, %s", bestSoFar, chainStr));
        }

        ReplayableMessage msg = messages.get();
        messages = messages.drop(1);

        return msg;
    }

}

File: src/main/ts/com/jcloisterzone/ai/player/RankingAiPlayer.ts
import type { GameStatePhaseReducer } from "../../game/GameStatePhaseReducer.js";
import { PortalCapability } from "../../game/capability/PortalCapability.js";
import type { Message } from "../../io/message/Message.js";
import { PlaceTileMessage } from "../../io/message/PlaceTileMessage.js";
import type { ReplayableMessage } from "../../io/message/ReplayableMessage.js";
import type { Player } from "../../Player.js";
import type { GameState } from "../../game/state/GameState.js";
import { AiPlayer } from "../AiPlayer.js";
import type { GameStateRanking } from "../GameStateRanking.js";

/** AI that searches all move-chains for the current player's turn-part and keeps the
 *  chain whose resulting state ranks highest. Returns the chain one message per call.
 *  Port of `ai/player/RankingAiPlayer`. */
export abstract class RankingAiPlayer extends AiPlayer {
  private readonly stateRanking: GameStateRanking;
  private readonly phaseReducer: GameStatePhaseReducer;
  private readonly me: Player;
  private messages: ReplayableMessage[] = [];

  constructor(phaseReducer: GameStatePhaseReducer, me: Player) {
    super();
    this.phaseReducer = phaseReducer;
    this.me = me;
    this.stateRanking = this.createStateRanking(me);
  }

  protected abstract createStateRanking(me: Player): GameStateRanking;

  override apply(state: GameState): ReplayableMessage {
    if (this.messages.length === 0) {
      let bestSoFar = Number.NEGATIVE_INFINITY;
      const queue: Array<[GameState, ReplayableMessage[]]> = [[state, []]];
      while (queue.length > 0) {
        const [itemState, chain] = queue.shift()!;
        for (const msg of this.getPossibleActions(itemState)) {
          const newChain = chain.concat(msg);
          const newState = this.phaseReducer.apply(itemState, msg as unknown as Message);
          let end =
            newState.getActivePlayer() === null ||
            newState.getActivePlayer()!.getIndex() !== this.me.getIndex() ||
            newState.getTurnPlayer()!.getIndex() !== state.getTurnPlayer()!.getIndex();
          if (
            !end &&
            msg instanceof PlaceTileMessage &&
            newState.getLastPlaced()!.getTile().hasModifier(PortalCapability.MAGIC_PORTAL)
          ) {
            // hack to avoid bad performance on a Portal tile: rank placement then rank
            // meeple placement separately (may miss a good on-tile meeple placement)
            end = true;
          }
          if (end) {
            const ranking = this.stateRanking.apply(newState);
            if (ranking > bestSoFar) {
              bestSoFar = ranking;
              this.messages = newChain;
            }
          } else {
            queue.push([newState, newChain]);
          }
        }
      }
    }
    const msg = this.messages[0];
    this.messages = this.messages.slice(1);
    return msg;
  }
}
